Sources/CNIOLinux/include/io_uring.h

```diff
@@ -0,0 +1,410 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2021 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+// from liburing project, local copy needed for builds on linux platforms without liburing installed
+// (liburing will be disabled then though)
+    
+#ifdef __linux__
+
+#ifdef C_NIO_LIBURING_UNAVAILABLE // only pull in these definitions if liburing isn't actually available
+
+typedef int __kernel_rwf_t;
+
+/* SPDX-License-Identifier: (GPL-2.0 WITH Linux-syscall-note) OR MIT */
+/*
+ * Header file for the io_uring interface.
+ *
+ * Copyright (C) 2019 Jens Axboe
+ * Copyright (C) 2019 Christoph Hellwig
+ */
+#ifndef LINUX_IO_URING_H
+#define LINUX_IO_URING_H
+
+#include <linux/fs.h>
+#include <linux/types.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/*
+ * IO submission data structure (Submission Queue Entry)
+ */
+struct io_uring_sqe {
```

Comment:
unfortunately, we won't be able to just copy header files. The header files need to exactly match the structures on the machine where the code is executed. Why do we need to vendor those files, don't they come with liburing?

Reply:
They do come with liburing, basically those are only used for when you don't have liburing available on your system (to allow the shims to compile), but then uring will be disabled in the resulting framework, so the actual struct layout doesn't matter really at all, it is just for compilation. I can stub all of it manually to remove the need for those two headers, but it was quite a bit needed, so thus I pulled them in.

Reply:
I see. I think we should probably do the stubbing, but I'm happy to delay this to later.

@Lukasa WDYT?

Another option would be to enable/disable Uring on the command-line at compile-time. Given that we get distributed with SwiftPM, users do the build anyway. So they could do `swift build -Xswiftc -DSWIFT_NIO_ENABLE_IO_URING` or something?

Reply:
I think we should stub it as well, but I'm again happy to delay until later. Maybe the best delay to avoid copyright concerns is to adopt the compile-time define to omit the imports if we don't need them.